### PR TITLE
fix(binder): ExprVisitor and ExprMutator should visit agg_call order_by and filter

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title. By missing these fields during iteration, correlated input refs were not handled and caused panic as mentioned in #4762.

Note that correlated input refs in the `filter` clause are now correctly detected, but the optimizer rule is not handling it yet. Unlike `input` or `order_by`, subexpr in `filter` are not extracted to a child `Project` node, thus need special treatment in decorrelation rules. This shall be fixed separately.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 1]
```diff
@@ -1042,18 +1042,6 @@ impl LogicalAgg {
...
-            .iter()
-            .chain(group_exprs.iter())
-            .any(|e| e.has_table_function())
-        {
-            return Err(ErrorCode::NotImplemented(
-                "Table functions in agg call or group by is not supported yet".to_string(),
-                3814.into(),
-            )
-            .into());
-        }
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:1055]: This check was redundant and too restrictive:
* When table function is an input argument of agg call, or part of group by, it is already rejected by the `Project` node builder.
* When table function is a sibling of agg call, we are able to handle it.

Refer to planner tests in `agg.yaml` for examples of both cases above.

File: src/frontend/planner_test/tests/testdata/array.yaml [hunk 2]
```diff
@@ -72,13 +72,14 @@
     └─BatchValues { rows: [[]] }
 - sql: |
     select array_cat(array[233], array[array[array[66]]]);
-  binder_error: "Bind error: unable to find least restrictive type between integer[] and integer[][][]"
+  binder_error: 'Bind error: unable to find least restrictive type between integer[]
```
Comment (xiangjinwu) [inline src/frontend/planner_test/tests/testdata/array.yaml:75]: Styling differences generated by `./risedev do-apply-planner-test`. Not sure why they were introduced by #5345

**Review by xxchan — APPROVED** [review]

**Comment by xxchan** [comment]
> Note that correlated input refs in the `filter` clause are now correctly detected, but the optimizer rule is not handling it yet. Unlike `input` or `order_by`, subexpr in `filter` are not extracted to a child `Project` node, thus need special treatment in decorrelation rules. This shall be fixed separately.

Is there a test case/issue for that?

**Comment by xiangjinwu** [comment]
> > Note that correlated input refs in the `filter` clause are now correctly detected, but the optimizer rule is not handling it yet. Unlike `input` or `order_by`, subexpr in `filter` are not extracted to a child `Project` node, thus need special treatment in decorrelation rules. This shall be fixed separately.
> 
> Is there a test case/issue for that?

Yes. It is the same issue but in comments https://github.com/risingwavelabs/risingwave/issues/4762#issuecomment-1228037768 . For this reason I am not marking it closed by this PR.

File: src/frontend/src/expr/agg_call.rs [hunk 3]
```diff
@@ -156,6 +156,22 @@ impl AggCall {
...
+        &mut self.order_by
+    }
+
+    pub fn filter(&self) -> &Condition {
+        &self.filter
+    }
+
+    pub fn filter_mut(&mut self) -> &mut Condition {
+        &mut self.filter
+    }
```
[the remark this answers is not in the document]
Comment (xiangjinwu) [reply src/frontend/src/expr/agg_call.rs:174]: Was thinking the same, and an alternative to simply expose `children_iter`/`children_mut` to return a chained iterator over inputs + order_by + filter. Not sure which is better as of now.

**Pull request merged**